jax_galsim/interpolatedimage.py

```diff
@@ -0,0 +1,1114 @@
+    _maxk_xint = x_interpolant.urange() / _uscale / scale
+
+    val = _draw_with_interpolant_kval(
+        kxi,
+        kyi,
+        kymin,  # this is not a bug! we need the minimum for the full periodic space
+        kymin,
+        arr,
+        k_interpolant,
+    )
+
+    msk = (jnp.abs(kxi) <= _maxk_xint) & (jnp.abs(kyi) <= _maxk_xint)
+    xint_val = x_interpolant._kval_noraise(kx) * x_interpolant._kval_noraise(ky)
+    return jnp.where(msk, val * xint_val * pfac, 0.0)
+
+
+@partial(jax.jit, static_argnames=("interp",))
+def _draw_with_interpolant_kval(kx, ky, kxmin, kymin, zp, interp):
+    orig_shape = kx.shape
+    kx = kx.ravel()
+    kxi = jnp.floor(kx - kxmin).astype(jnp.int32)
+    kxp = kxi + kxmin
+    nkx_2 = zp.shape[1] - 1
+    nkx = nkx_2 * 2
+
+    ky = ky.ravel()
+    kyi = jnp.floor(ky - kymin).astype(jnp.int32)
+    kyp = kyi + kymin
+    nky = zp.shape[0]
+
+    def _body_1d(i, args):
+        z, wky, kyind, kxi, nkx, nkx_2, kxp, zp = args
+
+        kxind = (kxi + i) % nkx
+        _kx = kx - (kxp + i)
+        wkx = interp._xval_noraise(_kx)
+
+        val = jnp.where(
+            kxind < nkx_2,
+            zp[(nky - kyind) % nky, nkx - kxind - nkx_2].conjugate(),
+            zp[kyind, kxind - nkx_2],
+        )
+        z += val * wkx * wky
+
+        return [z, wky, kyind, kxi, nkx, nkx_2, kxp, zp]
+
+    def _body(i, args):
+        z, kxi, kyi, nky, nkx, nkx_2, kxp, kyp, zp = args
+        kyind = (kyi + i) % nky
+        _ky = ky - (kyp + i)
+        wky = interp._xval_noraise(_ky)
+        z = jax.lax.fori_loop(
+            -interp.xrange,
+            interp.xrange + 1,
+            _body_1d,
+            [z, wky, kyind, kxi, nkx, nkx_2, kxp, zp],
+        )[0]
+        return [z, kxi, kyi, nky, nkx, nkx_2, kxp, kyp, zp]
+
+    z = jax.lax.fori_loop(
+        -interp.xrange,
+        interp.xrange + 1,
+        _body,
+        [jnp.zeros(kx.shape, dtype=complex), kxi, kyi, nky, nkx, nkx_2, kxp, kyp, zp],
+    )[0]
+    return z.reshape(orig_shape)
+
+
+@jax.jit
+def _flux_frac(a, x, y, cenx, ceny):
+    def _body(d, args):
+        res, a, dx, dy, cenx, ceny = args
+        msk = (jnp.abs(dx) <= d) & (jnp.abs(dx) <= d)
+
+        res = res.at[d].set(
+            jnp.sum(
+                jnp.where(
+                    msk,
+                    a,
+                    0.0,
+                )
+            )
+        )
+
+        return [res, a, dx, dy, cenx, ceny]
+
+    res = jnp.zeros(a.shape[0], dtype=float) - jnp.inf
+    return jax.lax.fori_loop(
+        0, a.shape[0], _body, [res, a, x - cenx, y - ceny, cenx, ceny]
+    )[0]
+
+
+@jax.jit
+def _calculate_size_containing_flux(image, thresh):
+    cenx, ceny = image.center.x, image.center.y
+    x, y = image.get_pixel_centers()
+    fluxes = _flux_frac(image.array, x, y, cenx, ceny)
+    msk = fluxes >= -jnp.inf
+    fluxes = jnp.where(msk, fluxes, jnp.max(fluxes))
+    d = jnp.arange(image.array.shape[0]) + 1.0
+    expfac = 4.0
```

Comment:
could you explain what is this quantity and why it's hardcoded?

Reply:
done!